### PR TITLE
[IMP] Website_slide: add slide count and time to slide categories

### PR DESCRIPTION
**Pull request by angv-odoo**
This commit intends to add information about the number of slides,
as well as the time needed to finish said slides to the slide sections.
We now also round completion time values up to minute when getting said values
from Youtube, Google drive or Vimeo.

Task-3054391

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/website_slides/models/slide_slide.py [hunk 1]
```diff
@@ -372,7 +372,7 @@ def _compute_embed_counts(self):
         for slide in self:
             slide.embed_count = mapped_data.get(slide.id, 0)
 
-    @api.depends('slide_ids.sequence', 'slide_ids.slide_category', 'slide_ids.is_published', 'slide_ids.is_category')
+    @api.depends('slide_ids.sequence', 'slide_ids.slide_category', 'slide_ids.is_published', 'slide_ids.is_category', 'slide_ids.completion_time', 'slide_ids.active')
```
Comment (thjo-odoo) [inline addons/website_slides/models/slide_slide.py:375]: I see that you added `slide_ids.active` to the dependencies but I don't see you use it.

Are the archived slides really removed from your computation ? 

Since, I think, if a slide is archived it should not be taken into account in the statistics but, when it is unarchived, then we add to the computation.

File: addons/website_slides/models/slide_slide.py [hunk 2]
```diff
@@ -1217,9 +1222,9 @@ def _fetch_google_drive_metadata(self, image_url_only=False):
                 slide_metadata['slide_type'] = 'google_drive_video'
 
         elif self.slide_category == 'video':
-            completion_time = float(
+            completion_time = round(float(
                 google_drive_values.get('videoMediaMetadata', {}).get('durationMillis', 0)
-                ) / (60 * 60 * 1000)  # millis to hours conversion
+                ) / (60 * 1000)) / 60  # millis to hours conversion with minute accuracy
```
Comment (thjo-odoo) [inline addons/website_slides/models/slide_slide.py:1227]: *rounded to the minute* would be a better description instead of accuracy

**Review by thjo-odoo — COMMENTED** [review]
Hello @angv-odoo,

Thanks for your work, I just have a suggestion and a question for this.

Otherwise it looks good to me.

Cheers !

File: addons/website_slides/models/slide_slide.py (hunk 1 above)
Comment (thjo-odoo) [reply addons/website_slides/models/slide_slide.py]: Just did a recheck on the runbot and I was wrong for this one indeed `slide_ids.active` is necessary for the frontend computation of a section's duration, my bad.

File: addons/website_slides/models/slide_slide.py [hunk 3]
```diff
@@ -393,14 +397,19 @@ def _compute_slides_statistics_category(self, read_group_res):
...
             slide_category = res_group.get('slide_category')
+            slide_completion_time = res_group.get('completion_time')
             if slide_category:
                 slide_category_count = res_group.get('__count', 0)
                 result[cid]['nbr_%s' % slide_category] = slide_category_count
                 result[cid]['total_slides'] += slide_category_count
+                result[cid]['completion_time'] += slide_completion_time
+        for record in self:
+            if not record.is_category:
+                result[record.id]['completion_time'] = record.completion_time
```
Comment (awa-odoo) [inline addons/website_slides/models/slide_slide.py:412]: "completion_time" is not a computed field, so it should not be set here?

File: addons/website_slides/models/slide_slide.py [hunk 4]
```diff
@@ -229,7 +229,7 @@ class Slide(models.Model):
     nbr_infographic = fields.Integer("Number of Images", compute='_compute_slides_statistics', store=True)
     nbr_article = fields.Integer("Number of Articles", compute='_compute_slides_statistics', store=True)
     nbr_quiz = fields.Integer("Number of Quizs", compute="_compute_slides_statistics", store=True)
-    total_slides = fields.Integer(compute='_compute_slides_statistics', store=True)
+    total_slides = fields.Integer(compute='_compute_slides_statistics', store=False, compute_sudo=True)
```
Comment (awa-odoo) [inline addons/website_slides/models/slide_slide.py:232]: Any reason to "unstore" this? It was probably stored for a reason (sorting, search, ...).

File: addons/website_slides/views/slide_slide_views.xml [hunk 5]
```diff
@@ -332,9 +332,13 @@
             <field name="arch" type="xml">
                 <search string="Search Contents">
                     <field name="name"/>
+                    <field name="channel_id"/>
+                    <field name="user_id"/>
+                    <field name="tag_ids"/>
```
Comment (awa-odoo) [inline addons/website_slides/views/slide_slide_views.xml:337]: Seems a bit unrelated as well, it could stay in this commit as it's a small change but it should be mentioned in the commit message.

**Review by awa-odoo — COMMENTED** [review]
Hello @angv-odoo 

A few questions on this one, as the code seems a bit far from the commit purpose.

Thanks & cheers!

File: addons/website_slides/models/slide_channel.py [hunk 6]
```diff
@@ -250,7 +250,7 @@ def _get_default_enroll_msg(self):
     total_slides = fields.Integer('Number of Contents', compute='_compute_slides_statistics', store=True)
     total_views = fields.Integer('Visits', compute='_compute_slides_statistics', store=True)
     total_votes = fields.Integer('Votes', compute='_compute_slides_statistics', store=True)
-    total_time = fields.Float('Duration', compute='_compute_slides_statistics', digits=(10, 2), store=True)
+    total_time = fields.Float('Duration', compute='_compute_slides_statistics', digits=(10, 7), store=True)
```
[the remark this answers is not in the document]
Comment (angv-odoo) [reply addons/website_slides/models/slide_channel.py:253]: Hi :) thanks for the review! We do need a change in precision here, as right now in master if you would want to save total_time value of let's say 00:10, then change it to 00:09 or 00:11, in all three cases the time show would be 00:10. 
The reason behind it is that we are saving information with 2 digit accuracy and it is not enough for the widget to calculate the minutes without error. It's the same with the "completion_time" value, in both cases I found 7 digit precision to be the lowest acceptable value where we can always have correct amount of minutes displayed.

File: addons/website_slides/models/slide_slide.py (hunk 4 above)
Comment (angv-odoo) [reply addons/website_slides/models/slide_slide.py:232]: So the change we made with completion time and total slides is the following: When you are an administrator, you see both published and unpublished slides, so total_slides should show you the value that is sum of all of these slides. For portal user, total_slides should calculate only published slides. The easiest way to do it that me and Florian could think of was to recalculate total_slides value every time user opens a page. This way value will be calculated every time page opens based on the level of access rights of a user, and at the same time whenever admin publishes or unpublished courses, portal user will get updated in comparatively "real time"

File: addons/website_slides/views/slide_slide_views.xml (hunk 5 above)
Comment (angv-odoo) [reply addons/website_slides/views/slide_slide_views.xml:337]: I'll update commit message with this info :+1:

File: addons/website_slides/models/slide_slide.py (hunk 3 above)
Comment (angv-odoo) [reply addons/website_slides/models/slide_slide.py:411]: "completion_time" is a weird case, it's not computed field for content, but we need to compute it when we are looking at the section record.

File: addons/website_slides/views/slide_slide_views.xml [hunk 7]
```diff
@@ -144,7 +144,7 @@
                                         <field name="user_id" string="Responsible" domain="[('share', '=', False)]" widget="many2one_avatar"/>
                                         <label for="completion_time"/>
                                         <div>
-                                            <field name="completion_time" widget="float_time" class="oe_inline"/>
+                                            <field name="completion_time"  class="oe_inline"/>
```
Comment (awa-odoo) [inline addons/website_slides/views/slide_slide_views.xml:147]: Doesn't that mess up the display and show `1,67 hours` ?

**Review by awa-odoo — COMMENTED** [review]
Hello @angv-odoo 

Sorry for the answer delay.
A few more comments, we are getting close on this one, thank you for your follow-up.

Cheers!

File: addons/website_slides/models/slide_slide.py [hunk 8]
```diff
@@ -396,14 +396,19 @@ def _compute_slides_statistics_category(self, read_group_res):
         """ Compute statistics based on all existing slide categories """
         slide_categories = self.env['slide.slide']._fields['slide_category'].get_values(self.env)
         keys = ['nbr_%s' % slide_category for slide_category in slide_categories]
-        result = dict((cid, dict((key, 0) for key in keys + ['total_slides'])) for cid in self.ids)
+        result = dict((cid, dict((key, 0) for key in keys + ['total_slides', 'completion_time'])) for cid in self.ids)
```
[the remark this answers is not in the document]
Comment (angv-odoo) [reply addons/website_slides/models/slide_slide.py]: Hey, I changed things around, and figured out how to update completion time to include  unpublished content for the users who have a right to see said content and publish them (the way LNA wanted before) by calling a compute_completion_time for a category whenever said category data is getting prepared to be rendered in a page. 
Do you think this is a right way of solving the issue?

[the remark this answers is not in the document]
Comment (angv-odoo) [reply addons/website_slides/models/slide_slide.py]: Just realized... Since we are not including unpublished content for the users in slide count and completion time values anymore, there are no more non-stored fields in that computed method, we are storing all the fields into database :sweat:

File: addons/website_slides/models/slide_slide.py [hunk 9]
```diff
@@ -375,16 +375,16 @@ def _compute_embed_counts(self):
         for slide in self:
             slide.embed_count = mapped_data.get(slide.id, 0)
 
-    @api.depends('slide_ids.sequence', 'slide_ids.slide_category', 'slide_ids.is_published', 'slide_ids.is_category')
+    @api.depends('slide_ids.sequence', 'slide_ids.slide_category', 'slide_ids.is_published', 'slide_ids.is_category', 'slide_ids.completion_time', 'slide_ids.active')
```
Comment (awa-odoo) [inline addons/website_slides/models/slide_slide.py:378]: Probably don't need to depend on "active" ?

File: addons/website_slides/models/slide_slide.py [hunk 10]
```diff
@@ -396,14 +396,18 @@ def _compute_slides_statistics_category(self, read_group_res):
...
             slide_category = res_group.get('slide_category')
             if slide_category:
                 slide_category_count = res_group.get('__count', 0)
-                result[cid]['nbr_%s' % slide_category] = slide_category_count
+                result[cid]['nbr_%s' % slide_category] += slide_category_count
                 result[cid]['total_slides'] += slide_category_count
+                result[cid]['completion_time'] += res_group['completion_time']
+        for record in self:
+            if not record.is_category:
+                result[record.id]['completion_time'] = record.completion_time
```
Comment (awa-odoo) [inline addons/website_slides/models/slide_slide.py:410]: Do we need this part?

This will just set the completion time on regular slides that already have it, which is strange.

File: addons/website_slides/models/slide_slide.py (hunk 9 above)
Comment (awa-odoo) [inline addons/website_slides/models/slide_slide.py:378]: The "completion_time" field should have this method in its definition then.

File: addons/website_slides/views/website_slides_templates_course.xml [hunk 11]
```diff
@@ -766,7 +775,7 @@
         </div>
         <div class="card-footer bg-white text-600">
             <div class="d-flex align-items-center small">
-                <span class="fw-bold me-auto" t-field="slide.completion_time" t-options='{"widget": "float_time"}'/>
+                <span class="fw-bold me-auto" t-field="category['category'].completion_time" t-options='{"widget": "float_time"}'/>
```
Comment (awa-odoo) [inline addons/website_slides/views/website_slides_templates_course.xml:778]: Hmm is this correct? Not sure why we use the category completion now instead of the slide completion for this part.

**Review by awa-odoo — COMMENTED** [review]
Hello @angv-odoo 

Some additional comments.
Can I ask you to focus on this one? Would like to close it today as we've been going back and forth for a while now.

Thanks & cheers!

File: addons/website_slides/models/slide_slide.py (hunk 10 above)
Comment (angv-odoo) [reply addons/website_slides/models/slide_slide.py:410]: Hey :) Because we are computing all of the statistics in the function (both category and not category), we need this part of code so that already computed info about slides isn't lost

File: addons/website_slides/models/slide_slide.py (hunk 9 above)
Comment (angv-odoo) [reply addons/website_slides/models/slide_slide.py:378]: True, corrected that part :smile:

Comment (angv-odoo) [reply addons/website_slides/models/slide_slide.py:378]: we do need the dependance on "active", it may not show when simply adding slides to the category, but if you delete one of the categories and then try to add new ones time computation will get all weird without it.

**Comment by awa-odoo** [comment]
@robodoo r+

**Comment by awa-odoo** [comment]
@robodoo r+

File: addons/website_slides/models/slide_slide.py (hunk 9 above)
Comment (angv-odoo) [reply addons/website_slides/models/slide_slide.py]: Hey :) Does it really need to have this method in the definition, considering that it is actually computed in multiple different places?  (What I mean is that category completion_time is calculated here, google_drive,vimeo and youtube completion_time are calculated in their respective functions, document completion_time is calculated in the separate function, etc?) 

The issue seems to be that if we assign this method to the completion time, then we also need to asign "recursive=True" value to it, as we are calculating the category completion time by summing up the other completion time values, and then we break the cascade deletion functionality on the course :smile:

Comment (angv-odoo) [reply addons/website_slides/models/slide_slide.py]: Same issue, separating methods didn't help it :confused:

**Comment by angv-odoo** [comment]
Hi @awa-odoo , all checks were successful :partying_face: Is there anything else we can do for this task?

**Comment by awa-odoo** [comment]
@robodoo r+
